closure iterator that captures the outer proc environment generates incompatible pointer assignments

This is an another pointer type compatibility issue, found while compiling testament.

### Example

```nim
proc main() =
  let x = 0
  iterator c(): int {.closure.} =
    yield x

  var iter = c

main()
```

### Actual Output

```
@mtest.nim.c: In function ‘c__test_3’:
@mtest.nim.c:231:20: error: assignment to ‘tyObject_Env_testdotnim_c___QRjx23R1ttwg6yg9a8vWR2w *’ from incompatible pointer type ‘tyObject_Env_testdotnim_main___lcRWa04p7qwyxS5SaoH5UA *’ [-Wincompatible-pointer-types]
  231 |         colonenvP_ = (tyObject_Env_testdotnim_main___lcRWa04p7qwyxS5SaoH5UA*) ClE_0;
      |                    ^
```

### Expected Output

Compilation should be successful

### References

- [Porting to GCC 14#Type checking on pointer types](https://gcc.gnu.org/gcc-14/porting_to.html#incompatible-pointer-types)

<!--
* related issues, PRs, discussions
* any additional or relevant links for people to better understand things and work with you
-->

<!--
check the issue for the following things:
* *easy to read*: lead with the primary information, follow with details, and is cordial in tone
* *easy to triage*: title is clear and short for others who know the language but not your issue
* *easy to reproduce*: example is minimal, with all necessary reproduction steps
* *clearly demonstrates the problem*: actual and expected outputs are minimal, with key information easy to spot
-->

--- a/compiler/backend/cgen.nim
+++ b/compiler/backend/cgen.nim
@@ -683,6 +683,7 @@ proc closureSetup(p: BProc, prc: PSym) =
   # prc.ast[paramsPos].last contains the type we're after:
   var ls = lastSon(prc.ast[paramsPos])
   p.config.internalAssert(ls.kind == nkSym, prc.info, "closure generation failed")
+  p.config.internalAssert(ls.typ == ls.sym.typ) # sanity check
   var env = ls.sym.position + 1 # parameters start at ID 1
 
   let n = newLocalRef(LocalId(env), ls.info, ls.typ)

--- a/compiler/sem/lambdalifting.nim
+++ b/compiler/sem/lambdalifting.nim
@@ -769,6 +769,8 @@ proc liftLambdas*(g: ModuleGraph; fn: PSym, body: PNode;
       param = getHiddenParam(g, fn)
     else:
       param.typ = t # replace with the correct type
+      # also update the symbol *node's* type
+      fn.ast[paramsPos][^1].typ = t
 
     prepareInnerRoutines(d, idgen, t, fn.info)
     # the environment instance is not setup here; that's done at the iterator's

